### DISCUSSION TITLE
Golang 如何进行类型检查 · /golang-typecheck

### DISCUSSION THREAD

**Discussion by draveness** [draveness/blog-comments#544] [Announcements]
https://draveness.me/golang-typecheck

**Comment by chinazhenzhen** [comment]
之前读过博主关于数据库的文章，感觉收获挺大的。再来翻看楼主的博客，发现博主又开始更新Go的博文了。不知道博主以后会不会更新自己对Go的interface的理解相关的博文。

**Comment by draveness** [comment]
> @chinazhenzhen
> 之前读过博主关于数据库的文章，感觉收获挺大的。再来翻看楼主的博客，发现博主又开始更新Go的博文了。不知道博主以后会不会更新自己对Go的interface的理解相关的博文。

会的，Go 语言相关的都会写

**Comment by payall4u** [comment]
> 只使用动态类型检查的编程语言就叫做动态类型编程语言，常见的动态类型编程语言就包括 Javascript、Ruby 和 PHP，这些编程语言在使用上非常灵活也不需要经过编译器的编译。

这里有点错误，还是需要经过编译器编译的。另外博主的文章很好，会看很多遍。

**Comment by draveness** [comment]
> 这里有点错误，还是需要经过编译器编译的。另外博主的文章很好，会看很多遍。

不过动态类型、静态类型这些概念本身就没有**标准的定义**，没法深究。我的理解是，大部分的动态类型编程语言是解释型的，也就是一边解释一边执行，没有生成中间代码或者二进制文件的过程。

**Comment by payall4u** [comment]
> > 这里有点错误，还是需要经过编译器编译的。另外博主的文章很好，会看很多遍。
>
> 不过动态类型、静态类型这些概念本身就没有标准的定义，没法深究。我的理解是，大部分的动态类型编程语言是解释型的，也就是一边解释一边执行，没有生成中间代码或者二进制文件的过程。

一部分解释执行的（比如python2，py3没看过）流程是：编译->ast->生成字节码->解释执行字节码。一部分动态类型语言的早起实现是基于ast的（比如ruby1.8-），也就是直接解释ast。不过现在多数使用解释字节码的方式。

**Comment by draveness** [comment]
> 一部分解释执行的（比如python2，py3没看过）流程是：编译->ast->生成字节码->解释执行字节码。一部分动态类型语言的早起实现是基于ast的（比如ruby1.8-），也就是直接解释ast。不过现在多数使用解释字节码的方式。

你这么理解是可以的，不过我觉得文中的说法也没有问题，所有的语言都需要经过**编译**这一过程，但是不一定有编译器(compiler)这东西，它们可能在解释器(interpreter)中进行这一过程。

**Comment by draveness** [comment]
> @chinazhenzhen
> 之前读过博主关于数据库的文章，感觉收获挺大的。再来翻看楼主的博客，发现博主又开始更新Go的博文了。不知道博主以后会不会更新自己对Go的interface的理解相关的博文。

接口相关的章节一节写完了，可以看一下 https://draveness.me/golang/docs/part2-foundation/ch04-basic/golang-interface/

**Comment by xiaopeifeng** [comment]
typo: 但是在类型检查阶段就会根据创建的类型将 make 替换成本的函数

---

*2020-06-05 UPDATES: 已修复*

**Comment by Formergg** [comment]
### typo

`NewSlice` 函数直接返回了一个 `TSLICE` 类型的结构体，~中~元素的类型信息 `r.Type` 也会存储在结构体中。

---

`NewSlice` 函数直接返回了一个 `TSLICE` 类型的结构体，元素的类型信息 `r.Type` 也会存储在结构体中。

---

*2020-07-13 UPDATES: 已修复*

**Comment by chenxinlong** [comment]
"还会在编程期引入类型信息" ，这里的 “编程期” 是个 typo 吗？

---

*2020-10-26 UPDATES: 已修复*

```diff --git a/design-and-implementation-of-go/content/docs/part1-prerequisite/ch02-compile/golang-typecheck.md b/design-and-implementation-of-go/content/docs/part1-prerequisite/ch02-compile/golang-typecheck.md
index 10341c8..8f65237 100644
--- a/design-and-implementation-of-go/content/docs/part1-prerequisite/ch02-compile/golang-typecheck.md
+++ b/design-and-implementation-of-go/content/docs/part1-prerequisite/ch02-compile/golang-typecheck.md
@@ -58,7 +58,7 @@ 静态类型检查和动态类型检查其实并不是两种完全冲突和对

 ## 2.3.3 执行过程

-Go 语言的编译器不仅使用静态类型检查来保证程序运行的类型安全，还会在编程期引入类型信息，让工程师能够使用反射来判断参数和变量的类型，当我们想要将 `interface`
转换成具体类型时就会进行动态类型检查，如果无法发生转换就可能会造成程序崩溃。
+Go 语言的编译器不仅使用静态类型检查来保证程序运行的类型安全，还会在编程期间引入类型信息，让工程师能够使用反射来判断参数和变量的类型，当我们想要将 `interface` 转换成具体类型时就会进行动态类型检查，如果无法发生转换就可能会造成程序崩溃。

```

**Comment by draveness** [comment]
> @chenxinlong "还会在编程期引入类型信息" ，这里的 “编程期” 是个 typo 吗？

其实不是，你可以理解成编译期间，但是我还是改一下

**Comment by sunnystitch** [comment]
想问个问题  这些都是博主从头到位看了go的源码写的么  像更多的看底层 看了博主博客想进一步了解底层 但是不知道怎么下手，有好的思路，比如从那块开始入手一步步循序渐进的把go底层给搞了

**Comment by dillanzhou** [comment]
虽然typecheck确实是类型检查的意思，但是感觉这里的类型检查是在检查和处理关键词类型而不是变量类型，和上面讨论的强弱类型、静态动态类型似乎不是一回事啊。

**Comment by Bingmang** [comment]
### typo
`cmd/compile/internal/types.NewSlice` 直。接返回了一个 TSLICE 类型的结构体

这里多了一个句号

---

*2021-04-27 UPDATES: 已修复*

**Comment by sch-git** [comment]
多了个太多：那么在概念上较真本身是没有太多太多实际价值的。
另外，博主的文章写的很好，很有深度，推荐给其他朋友了！

**Comment by Ray-Eldath** [comment]
我觉得没有 UB 就是严格的 strong typing 的定义吧，可以说 staticly strong typing = soundness...?  

当然也没法完全做到 sound。。语言被爆出 soundness hole 其实也是很正常的事情，只是感觉如果语言承诺并尽量修复（而不是像某些语言一样理直气壮【这就是 UB！】）那就是 strongly typed 得（比如 Java 和 Scala 就有一个藏了十多年的 hole）。总之我自己的理解是这样。


顺便如果在这个框里打英文光标会直接飘到左上角搜索框，必须在别的地方打完英文复制过来才行。。有点难受，感觉是个 bug？

最后感谢博主的文章，每次遇到不太了解的概念就会来看博主写的东西。我也是看了博主的文章才萌生想去看 gc 的想法，感谢博主。

**Comment by iam153** [comment]
对于不同类型的Node，怎么知道它的Left和Right分别对应哪个token呢？

**Comment by Bit-urd** [comment]
牛逼，老哥。太牛逼了，真的可以出本书了，大半夜两点多看的我睡不着了。

**Comment by alomerry** [comment]
博主我看 go1.20 里已经没有文中描述的类型检查了，而是走了 [types2 包中的代码进行检查了](https://github.com/golang/go/blob/go1.20.1/src/cmd/compile/internal/noder/noder.go#L82)，会考虑更新文章吗 :pray:

**Reply by draveness** [reply]
等到后面会整体改一下
